### DISCUSSION TITLE
Provide a sketch for the UI how to channel/use a exception message within an error callback  method.

### DISCUSSION THREAD

**Discussion by bigopensky** [mr-highball/ezthreads#19] [General]
Hi I try to understand your UI approach and want to implement  a call  that either succeeds or stops with an error using Linux an FreePascal 3.0.0 and GTK from the Lazarus UI.

In both cases I want to inform the user what is happening via dialog boxes.  But run into problems.

###  Setup the thread

I cannot setup the Thread in the form shown in the example

```delphi
function TMultiFileDownload.GetTotalBytes(const ATotalVarName: String): IEZThread;
...
  Result := NewEZThread;
  Result
    .AddArg('count', memo_filelist.Lines.Count) //used to determine file count in thread method
    .AddArg('totalName', ATotalVarName)
    .Setup(GetBytes, nil, nil); //setup the call to get sum of content lengths                                       
```

but have to use  the address operator @ because of `Error: Wrong number of parameters specified.`

```delphi
procedure TFrmMain.DoFileStats(const aJobIdent, aSrcDir: string);  
...
     Result := NewEZThread;
     Result.AddArg('FILES', 0)
           .AddArg('SIZE',  0)
           .AddArg('JOB.ID',  aJobIdent)
           .AddArg('SRC.DIR', aSrcDir)
           .Setup(@TaskCreateFileStats,
                  @TaskErrorFileStats,
                  @TaskSuccessFileStats);
     Result.Start;
```
### How to implement the error handler?

I can throw an error but don't know how to retrieve the message from the exception

```delphi
procedure TFrmMain.TaskErrorFileStats(const AThread : IEZThread);
...
    LSrcDir:=AThread['SRC.DIR'];
    if not DirectoryExists(LSrcDir) then begin
      LHelp:= Format('Unknown directory ''%s''',[LSrcDir]);
     //// Work around  
      AThread.AddArg('ERROR', LHelp);
      raise Exception.Create(LHelp);
    end;
```

### How to call the GUI elements in a right way 

When I call the MessageDlg `MessageDlg(LHelp, mtError, [mbOK], 0); `  my system crashes with a runtime error and General protection fault in `g_main_context_iteration(g_main_context_default, True);` when I'm in debug mode. Occurs this  because I'm already in the thread context? 

## Question?

Can you provide a hint or a sketch how to setup an simple UI example? 

## Code fragment

[CalcCheckSum.zip](https://github.com/mr-highball/ezthreads/files/13349715/CalcCheckSum.zip)

```delphi
unit UCalcChecksumFiles;

{$MODE OBJFPC}
{$LONGSTRINGS ON}

{$DEFINE EZTHREAD_TRACE}


interface

uses
  Classes, SysUtils, Forms, Controls, Graphics, Dialogs, StdCtrls,
  ExtCtrls, ComCtrls, Buttons, ezthreads;

type

  { TFrmMain }

  TFrmMain = class(TForm)
    BtnStop: TBitBtn;
    BtnRun: TBitBtn;
    BtnQuit: TBitBtn;
    EdtDirectory: TEdit;
    GbxTop: TGroupBox;
    procedure BtnQuitClick(Sender: TObject);
    procedure BtnRunClick(Sender: TObject);
    procedure BtnStopClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    FCurThread: IEZThread;
    procedure  DoFileStats(const aJobIdent, aSrcDir: string);
    procedure TaskCreateFileStats(const AThread : IEZThread);
    procedure TaskErrorFileStats(const AThread : IEZThread);
    procedure TaskSuccessFileStats(const AThread : IEZThread);
  public

  end;

var
  FrmMain: TFrmMain;

implementation

procedure TFrmMain.FormCreate(Sender: TObject);
begin
  // Init somthing
end;

// Trigger the task setup
procedure TFrmMain.BtnRunClick(Sender: TObject);
begin
  DoFileStats('FILE.STATS', EdtDirectory.Text);
end;

// Stop the task
procedure TFrmMain.BtnStopClick(Sender: TObject);
begin
  if Assigned(FCurThread) then FCurThread.Stop;
end;

// Close the app
procedure TFrmMain.BtnQuitClick(Sender: TObject);
begin
  // Clean up fCutTask here?
  Close;
end;

// Setup the task
procedure TFrmMain.DoFileStats(const aJobIdent, aSrcDir: string);
begin
     FCurThread := NewEZThread;
     FCurThread
           .AddArg('ERROR','')
           .AddArg('FILES', 0)
           .AddArg('SIZE',  0)
           .AddArg('JOB.ID',  aJobIdent)
           .AddArg('SRC.DIR', aSrcDir)
           .Setup(@TaskCreateFileStats,
                  @TaskErrorFileStats,
                  @TaskSuccessFileStats);
     FCurThread.Start;
end;

// Run the task content
procedure TFrmMain.TaskCreateFileStats(const AThread : IEZThread);
var
  LSrcDir: string;
  LHelp:   string;
begin
    EdtDirectory.Enabled:=FALSE;
    BtnQuit.Enabled:=FALSE;
    BtnRun.Enabled:=TRUE;
    BtnStop.Enabled:=TRUE;

    LSrcDir:=AThread['SRC.DIR'];
    if not DirectoryExists(LSrcDir) then begin
      LHelp:= Format('Unknown directory ''%s''',[LSrcDir]);
      AThread.AddArg('ERROR', LHelp);
      raise Exception.Create(LHelp);
    end;

    // Calculate the file stats by I set
    // dummy sizes and counts here
    AThread.AddArg('FILE',  10);
    AThread.AddArg('SIZE', 100);
end;

// Handle the error case
procedure TFrmMain.TaskErrorFileStats(const AThread : IEZThread);
var
   LHelp: string;
begin
    LHelp:= AThread['ERROR'];
    if lHelp = '' then exit;
    MessageDlg(LHelp, mtError, [mbOK], 0);
    EdtDirectory.Enabled:=TRUE;

    BtnRun.Enabled:=TRUE;
    BtnQuit.Enabled:=TRUE;
    BtnStop.Enabled:=FALSE;
    // Clean up fCutTask here?
end;

// Handle the success case
procedure TFrmMain.TaskSuccessFileStats(const AThread : IEZThread);
var
   lHelp:  string;
   lSize:  integer;
   lCount: integer;
begin
   lHelp  := AThread['JOB.ID'];
   lSize  := AThread['SIZE'];
   lCount := AThread['FILES'];
   MessageDlg('Thread '+LHelp+' finished!'+
              ' FILES: '+IntToStr(lCount)+
              ' SIZE: '+IntToStr(lSize),
              mtInformation, [mbOK], 0);
   // Clean up fCutTask here?
end;

{$R *.lfm}

end.
                       
```

**Comment by mr-highball** [comment]
Hi @bigopensky -

The first thing I noticed was that you're using _objfpc_ mode rather than _delphi_ mode that I normally use (there's a subtle difference that should be the reason why you have to use the `@` instead of how my examples show).

This is of course preference but I find the Delphi mode to be more forgiving...

Also I see you're referencing a bit more complicated example, but I also have a simpler UI example here

https://github.com/mr-highball/ezthreads/blob/a0a391120ba14dd4eb6b3983e874f29336227230/test/main.pas#L57C23-L57C23

Take note around the      UpdateSynchronizeStopEvents(True) as this is necessary for _events_ to be invoked in the UI thread's context. I'm not at my PC right now to look over you're example but will try to poke around tomorrow to give a better look

**Comment by bigopensky** [comment]
Hi, I have updated the code fragments and inserted `UpdateSynchronizeStopEvents(TRUE) `.  But had no success.

 I have added `Message` and `QueueAsyncCall` driven updates for the GUI. Both approaches work as proposed by FreePascal.  

At the moment I am coding hard an error message into the dictionary `L:247: AThread.AddArg(KEY_ERROR, LHelp)`  to get message out of the task into the GUI. 

[CalcCheckSum-V1.zip](https://github.com/mr-highball/ezthreads/files/13376522/CalcCheckSum-V1.zip)

**Reply by mr-highball** [reply]
hmm, so I downloaded the source and clicked on the "run" button (making sure to not have the directory present) and the error populated
![image](https://github.com/mr-highball/ezthreads/assets/35942565/bf2f01a7-4aec-4c63-ae7a-440a6bf6bf11)

from your previous post it seems like that's the _intended_ behavior, but why your system crashes a bit of a mystery to me right now.

it may come down to a _version of fpc_ thing...
![image](https://github.com/mr-highball/ezthreads/assets/35942565/a34ff18f-9c4b-492a-a9e5-7ec3a5e9c1c8)

if you open the UI tester does it run properly or crash the system too?
https://github.com/mr-highball/ezthreads/tree/master/test

**Reply by bigopensky** [reply]
OK it seems to run as intended under a Windows environment.

When I run the same tool with the pure `UpdateSynchronizeStopEvents(TRUE)` , Linux with an GTK2 backend approach, the application **crashes**.

![Screenshot_2023-11-17_02-44-43](https://github.com/mr-highball/ezthreads/assets/11876970/3f0fd190-1c02-4823-98c4-da5cb3addb7b)

## Test ezthreads_tester_ui

The UI tester crashes also within a Linux / GTK2 environment. 

![Screenshot_2023-11-17_03-20-59](https://github.com/mr-highball/ezthreads/assets/11876970/ece8df2a-83ba-482a-b994-25ca4bcbc30e)

It seem to me that the direct access to an GUI from the thread `main.pas.LINE:77-LINE:80`

```pascal
  LMsg := AThread['text'];
  edit_test.Text := LMsg;
  btn_edit_doit.Enabled := True;     
```
is problematic. I guess the program calls the update within the `LTHread....Events.UpdateOnStop(...)` scope. 

## Environment

```bash
~/Source/IfGDV-Tests/Test-Async/ezthreads/demo/ChecksumFiles$ cat /etc/lsb-release
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=18.04
DISTRIB_CODENAME=bionic
DISTRIB_DESCRIPTION="Ubuntu 18.04.6 LTS"

~/Source/IfGDV-Tests/Test-Async/ezthreads/demo/ChecksumFiles$ uname -a
Linux zulu 5.4.0-150-generic #167~18.04.1-Ubuntu SMP Wed May 24 00:51:42 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```
![Screenshot_2023-11-17_02-48-48](https://github.com/mr-highball/ezthreads/assets/11876970/9bb69c44-2ceb-457f-9cfc-9121cc093fc5)

**Reply by mr-highball** [reply]
Ahh I think you might just need to include cthreads in your project file (note the error message your screenshot shows)

https://wiki.freepascal.org/Multithreaded_Application_Tutorial

![Screenshot_20231116_214038_Brave.jpg](https://github.com/mr-highball/ezthreads/assets/35942565/d1c17c2d-d91e-4fd0-a35a-b7c42a0ed646)

**Reply by mr-highball** [reply]
Actually if that was from the Tester the cthreads inclusion should already be there... I think I'll have to just get a Linux environment and try to troubleshoot

**Reply by bigopensky** [reply]
## Code Fragment CalcCheckSum

Yes both `cthreads` and `cmem` as suggested by FPC are already invoked in the `CalcCheckSum` case.

```pascal
program CalcCheckSum;

{$mode objfpc}{$H+}

uses
  {$IFDEF UNIX}
  cthreads,
  cmem,
  {$ENDIF}
  {$IFDEF HASAMIGA}
  athreads,
  {$ENDIF}
  Interfaces, // this includes the LCL widgetset
  Forms, UCalcChecksumFiles, ezthreads
  { you can add units after this };

{$R *.res}

begin
  RequireDerivedFormResource:=True;
  Application.Scaled:=True;
  Application.Initialize;
  Application.CreateForm(TFrmMain, FrmMain);
  Application.Run;
end.                  
```

## UI Test case 

In the your UI tester case `ctheads` is blanked by `{$IFDEF UseCThreads}` ...

```pascal
program ezthreads_tester_ui;

{$mode objfpc}{$H+}

uses
  {$IFDEF UNIX}{$IFDEF UseCThreads}
  cthreads,
  {$ENDIF}{$ENDIF}
  Interfaces, // this includes the LCL widgetset
  Forms, main
  { you can add units after this };
...
```

When I change this 

```pascal
program ezthreads_tester_ui;

{$mode objfpc}{$H+}

uses
  {$IFDEF UNIX}
  cthreads,
  cmem,
  {$ENDIF}
  Interfaces, // this includes the LCL widgetset
  Forms, main
  { you can add units after this };            
```

it runs ...
